Core/Util/GameManager.cpp

```diff
@@ -108,94 +109,207 @@ bool GameManager::Uninstall(std::string name) {
+				isoFileIndex = i;
+			}
 		}
 	}
 
-	if (!isPSP) {
+	info->stripChars = stripChars;
+	info->numFiles = numFiles;
+	info->isoFileIndex = isoFileIndex;
+	// If a ZIP is detected as both, let's let the memstick game interpretation prevail.
+	if (isPSPMemstickGame) {
+		return ZipFileContents::PSP_GAME_DIR;
+	} else if (isZippedISO) {
+		return ZipFileContents::ISO_FILE;
+	} else {
+		return ZipFileContents::UNKNOWN;
+	}
+}
+
+bool GameManager::InstallGame(std::string url, std::string fileName, bool deleteAfter) {
+	if (installInProgress_) {
+		ERROR_LOG(HLE, "Cannot have two installs in progress at the same time");
+		return false;
+	}
+
+	if (!File::Exists(fileName)) {
+		ERROR_LOG(HLE, "Game file %s doesn't exist", fileName.c_str());
+		return false;
+	}
+
+	// Examine the URL to guess out what we're installing.
+	if (endsWithNoCase(url, ".cso") || endsWithNoCase(url, ".iso")) {
+		// It's a raw ISO or CSO file. We just copy it to the destination.
+		std::string shortFilename = GetFilenameFromPath(url);
+		return InstallRawISO(fileName, shortFilename, deleteAfter);
+	}
+
+	I18NCategory *sy = GetI18NCategory("System");
+	installInProgress_ = true;
+
+	std::string pspGame = GetSysDirectory(DIRECTORY_GAME);
+	INFO_LOG(HLE, "Installing '%s' into '%s'", fileName.c_str(), pspGame.c_str());
+	int error = 0;
+#ifdef _WIN32
+	struct zip *z = zip_open(ConvertUTF8ToWString(fileName).c_str(), 0, &error);
+#else
+	struct zip *z = zip_open(fileName.c_str(), 0, &error);
+#endif
+	if (!z) {
+		ERROR_LOG(HLE, "Failed to open ZIP file %s, error code=%i", fileName.c_str(), error);
+		return false;
+	}
+
+	ZipFileInfo info;
+	ZipFileContents contents = DetectZipFileContents(z, &info);
+	switch (contents) {
+	case ZipFileContents::PSP_GAME_DIR:
+		// InstallMemstickGame contains code to close z.
+		return InstallMemstickGame(z, fileName, pspGame, info.numFiles, info.stripChars, deleteAfter);
+	case ZipFileContents::ISO_FILE:
+		return InstallZippedISO(z, info.isoFileIndex, fileName, deleteAfter);
+	default:
 		ERROR_LOG(HLE, "File not a PSP game, no EBOOT.PBP found.");
 		installProgress_ = 0.0f;
 		installInProgress_ = false;
 		installError_ = sy->T("Not a PSP game");
 		InstallDone();
 		if (deleteAfter)
-			File::Delete(zipfile);
+			File::Delete(fileName);
+		return false;
+	}
+}
+
+bool GameManager::ExtractFile(struct zip *z, int file_index, std::string outFilename, size_t *bytesCopied, size_t allBytes) {
+	struct zip_stat zstat;
+	zip_stat_index(z, file_index, 0, &zstat);
+	size_t size = zstat.size;
+
+	// Don't spam the log.
+	if (file_index < 10) {
+		INFO_LOG(HLE, "Writing %d bytes to '%s'", (int)size, outFilename.c_str());
+	}
+
+	zip_file *zf = zip_fopen_index(z, file_index, 0);
+	FILE *f = File::OpenCFile(outFilename, "wb");
+	if (f) {
+		size_t pos = 0;
+		const size_t blockSize = 1024 * 128;
+		u8 *buffer = new u8[blockSize];
+		while (pos < size) {
+			size_t bs = std::min(blockSize, size - pos);
+			zip_fread(zf, buffer, bs);
```

Comment:
Should we check the return in case of zip file corruption?  I guess we can probably assume that's unlikely...

-[Unknown]

Reply:
Yup, we should. Done